File: video-agent-test/internal/prompt/prompt.go
package prompt

import (
	"fmt"
	"os"
	"path/filepath"
	"strings"

	"gopkg.in/yaml.v3"
)

type Document struct {
	Path        string
	Frontmatter map[string]any
	Body        string
}

func ParseFile(path string) (*Document, error) {
	absPath, err := filepath.Abs(path)
	if err != nil {
		return nil, err
	}

	raw, err := os.ReadFile(absPath)
	if err != nil {
		return nil, err
	}

	content := string(raw)
	if !strings.HasPrefix(content, "---\n") {
		return nil, fmt.Errorf("missing YAML frontmatter")
	}

	parts := strings.SplitN(content, "---\n", 3)
	if len(parts) < 3 {
		return nil, fmt.Errorf("invalid YAML frontmatter")
	}

	frontmatter := map[string]any{}
	if strings.TrimSpace(parts[1]) != "" {
		if err := yaml.Unmarshal([]byte(parts[1]), &frontmatter); err != nil {
			return nil, fmt.Errorf("YAML parse failed: %w", err)
		}
	}

	body := strings.TrimSpace(parts[2])
	if body == "" {
		return nil, fmt.Errorf("prompt body is empty")
	}

	return &Document{
		Path:        absPath,
		Frontmatter: frontmatter,
		Body:        body,
	}, nil
}

File: agent/packages/opencode/src/video/prompt.ts
import * as fs from "fs/promises"
import * as path from "path"
import matter from "gray-matter"

export interface PromptDocument {
  path: string
  frontmatter: Record<string, unknown>
  body: string
}

export async function parsePromptFile(filePath: string): Promise<PromptDocument> {
  const absPath = path.resolve(filePath)
  const raw = await fs.readFile(absPath, "utf8")
  if (!raw.startsWith("---")) throw new Error("missing YAML frontmatter")

  const parsed = matter(raw)
  const frontmatter =
    parsed.data && typeof parsed.data === "object" && !Array.isArray(parsed.data)
      ? (parsed.data as Record<string, unknown>)
      : {}
  const rawFrontmatter = raw.match(/^---\r?\n([\s\S]*?)\r?\n---/)?.[1]
  const rawRatio = rawFrontmatter?.match(/^ratio:\s*("?)([^"\n#]+)\1\s*(?:#.*)?$/m)?.[2]?.trim()
  if (rawRatio) frontmatter.ratio = rawRatio
  const body = parsed.content.trim()
  if (!body) throw new Error("prompt body is empty")

  return {
    path: absPath,
    frontmatter,
    body,
  }
}
